File: samples/dnn/multi_object_tracker.py
#!/usr/bin/env python
'''
Follow several people at once with cv.MultiTracker.

A detector finds people in each frame and the tracker decides which box belongs to which person,
so every box keeps its number for as long as that person is on screen.

Copyright (C) 2026, BigVision LLC.

How to use:
    Download the models once:
        python download_models.py multi_object_tracker

    Then run:
        python multi_object_tracker.py --input=vtest.avi
        python multi_object_tracker.py --input=vtest.avi --reid

    --reid also loads a re-identification network, which helps ids survive when people cross.

    Set OPENCV_DOWNLOAD_CACHE_DIR to where the models were downloaded, and
    OPENCV_SAMPLES_DATA_PATH to opencv/samples/data.
'''
import argparse
import os.path
import numpy as np
import cv2 as cv

from common import *

def get_args_parser():
    backends = ("default", "openvino", "opencv", "vkcom", "cuda")
    targets = ("cpu", "opencl", "opencl_fp16", "ncs2_vpu", "hddl_vpu", "vulkan", "cuda", "cuda_fp16")

    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('--zoo', default=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'models.yml'),
                        help='An optional path to file with preprocessing parameters.')
    parser.add_argument('--input', '-i', default='vtest.avi', help='Path to video file or camera index.')
    parser.add_argument('--reid', action='store_true',
                        help='Also use appearance, so ids survive when people cross.')
    parser.add_argument('--conf', type=float, default=0.25, help='Detector confidence threshold.')
    parser.add_argument('--nms', type=float, default=0.45, help='Detector NMS threshold.')
    parser.add_argument('--backend', default="default", type=str, choices=backends,
                        help="Choose one of computation backends.")
    parser.add_argument('--target', default="cpu", type=str, choices=targets,
                        help="Choose one of target computation devices.")
    args, _ = parser.parse_known_args()
    add_preproc_args(args.zoo, parser, 'multi_object_tracker', prefix="", alias="multi_object_tracker")
    add_preproc_args(args.zoo, parser, 'multi_object_tracker', prefix="yolo_", alias="multi_object_tracker")
    parser = argparse.ArgumentParser(parents=[parser],
                                     description='Multi-object tracking using OpenCV.',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    return parser.parse_args()


def color_for_id(track_id):
    '''One colour per id, so a person keeps the same colour while they are tracked.'''
    rng = np.random.RandomState(track_id * 7919 % 2**31)
    return tuple(int(c) for c in rng.randint(64, 256, 3))


def detect(frame, net, size, scale, swap_rb, conf_threshold, nms_threshold):
    '''
    yolov8 emits one [1, 84, 8400] tensor: four box numbers then one score per class, for every
    candidate. The frame is padded to a square first so nothing is distorted, which is why the
    boxes are scaled back by the padded length afterwards.
    '''
    height, width = frame.shape[:2]
    length = max(height, width)
    square = np.zeros((length, length, 3), dtype=frame.dtype)
    square[:height, :width] = frame
    back_scale = length / float(size)

    blob = cv.dnn.blobFromImage(square, scale, (size, size), swapRB=swap_rb, crop=False)
    net.setInput(blob)
    out = net.forward()
    candidates = out.reshape(out.shape[1], -1).T        # [8400, 84]

    boxes, scores = [], []
    for row in candidates:
        class_scores = row[4:]
        best = int(np.argmax(class_scores))
        if class_scores[best] < conf_threshold or best != 0:   # class 0 is "person" in COCO
            continue
        cx, cy, w, h = row[:4]
        boxes.append([float(cx - w / 2), float(cy - h / 2), float(w), float(h)])
        scores.append(float(class_scores[best]))

    if not boxes:
        return [], [], []

    keep = np.array(cv.dnn.NMSBoxes(boxes, scores, conf_threshold, nms_threshold)).flatten()
    out_boxes = [tuple(v * back_scale for v in boxes[i]) for i in keep]
    return out_boxes, [scores[i] for i in keep], [0] * len(keep)


def describe(frame, reid, boxes):
    '''One L2-normalised descriptor per detection, which is what the appearance overload wants.'''
    if len(boxes) == 0:
        return None
    height, width = frame.shape[:2]
    rows = []
    for x, y, w, h in boxes:
        x0, y0 = max(0, int(x)), max(0, int(y))
        x1, y1 = min(width, int(x + w)), min(height, int(y + h))
        if x1 - x0 < 2 or y1 - y0 < 2:
            x0, y0, x1, y1 = 0, 0, min(2, width), min(2, height)
        feature = reid.predict(frame[y0:y1, x0:x1])[0].reshape(-1).astype(np.float32)
        norm = np.linalg.norm(feature)
        if norm > 0:
            feature = feature / norm
        rows.append(feature)
    return np.array(rows, dtype=np.float32)


def main():
    args = get_args_parser()

    yolo = cv.dnn.readNetFromONNX(findModel(args.yolo_model, args.yolo_sha1))
    yolo.setPreferableBackend(get_backend_id(args.backend))
    yolo.setPreferableTarget(get_target_id(args.target))

    reid = None
    if args.reid:
        reid = cv.dnn.Model(cv.dnn.readNetFromONNX(findModel(args.model, args.sha1)))
        # (pixel - mean) * scale, per channel, with the std folded into the scale. swapRB swaps
        # channel 0 and 2 of both, so each is given in the model's own RGB order.
        reid.setInputMean(args.mean)
        reid.setInputScale([args.scale / s for s in args.std])
        reid.setInputSize(args.width, args.height)
        reid.setInputSwapRB(args.rgb)
        reid.setPreferableBackend(get_backend_id(args.backend))
        reid.setPreferableTarget(get_target_id(args.target))

    params = cv.MultiTracker_Params()
    if args.reid:
        params.embeddingWeight = 0.4    # appearance helps, but motion still leads
    tracker = cv.MultiTracker.create(params)

    source = args.input if not str(args.input).isdigit() else int(args.input)
    if isinstance(source, str):
        source = findFile(source)
    cap = cv.VideoCapture(source)
    if not cap.isOpened():
        print('Could not open the input', args.input)
        return

    win_name = 'MULTI OBJECT TRACKING'
    cv.namedWindow(win_name, cv.WINDOW_NORMAL)
    font_face = cv.FontFace("sans")
    std_size, std_weight, std_img_size = 15, 400, 512

    while True:
        ok, frame = cap.read()
        if not ok:
            break

        font_size = max(10, int(std_size * frame.shape[1] / std_img_size))
        font_weight = max(200, int(std_weight * frame.shape[1] / std_img_size))

        boxes, scores, classes = detect(frame, yolo, args.yolo_width, args.yolo_scale,
                                        args.yolo_rgb, args.conf, args.nms)

        if args.reid:
            features = describe(frame, reid, boxes)
            ids, out_boxes, out_classes = tracker.update(boxes, scores, classes, features)
        else:
            ids, out_boxes, out_classes = tracker.update(boxes, scores, classes)

        for track_id, box in zip(ids, out_boxes):
            color = color_for_id(int(track_id))
            x, y, w, h = (int(v) for v in box)
            cv.rectangle(frame, (x, y, w, h), color, 2)

            label = str(track_id)
            badge_w = font_size * len(label) + 10
            cv.rectangle(frame, (x, max(0, y - font_size - 6), badge_w, font_size + 6), color, cv.FILLED)
            # FontFace overload: (img, text, org, color, fface, size, weight)
            cv.putText(frame, label, (x + 4, max(font_size, y - 4)), (0, 0, 0),
                       font_face, font_size, font_weight)

        status = 'tracking %d  (detections %d)%s' % (len(ids), len(boxes),
                                                     '  [re-id on]' if args.reid else '')
        cv.putText(frame, status, (10, font_size + 5), (0, 255, 0),
                   font_face, font_size, font_weight)

        cv.imshow(win_name, frame)
        if cv.waitKey(30) >= 0:
            break


if __name__ == '__main__':
    main()

File: samples/dnn/multi_object_tracker.cpp
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
// Copyright (C) 2026, BigVision LLC, all rights reserved.
// Third party copyrights are property of their respective owners.

#include <iostream>
#include <map>
#include <vector>

#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/video/tracking.hpp>

#include "common.hpp"

using namespace std;
using namespace cv;
using namespace cv::dnn;

const string about =
    "Follow several people at once with cv::MultiTracker.\n\n"
    "A detector finds people in each frame and the tracker decides which box belongs to which\n"
    "person, so every box keeps its number for as long as that person is on screen.\n\n"
    "Usage:\n"
    "  example_dnn_multi_object_tracker multi_object_tracker\n"
    "  example_dnn_multi_object_tracker multi_object_tracker --input=vtest.avi\n"
    "  example_dnn_multi_object_tracker multi_object_tracker --input=vtest.avi --reid\n\n"
    "--reid also loads a re-identification network, which helps ids survive when people cross.\n";

const string keys =
    "{ help  h        |                   | show this message }"
    "{ @alias         |                   | model alias from models.yml }"
    "{ zoo            | models.yml        | model description file }"
    "{ input i        | vtest.avi         | video file or camera index }"
    "{ reid           |                   | also use appearance to keep ids apart }"
    "{ conf           | 0.25              | detector confidence threshold }"
    "{ nms            | 0.45              | detector NMS threshold }"
    "{ device         | 0                 | camera index when --input is not a file }"
    "{ backend        | 0                 | 0 automatic, 1 Halide, 2 OpenVINO, 3 OpenCV, 4 VKCOM, 5 CUDA }"
    "{ target         | 0                 | 0 CPU, 1 OpenCL, 2 OpenCL FP16, 3 Myriad, 4 Vulkan, 6 CUDA, 7 CUDA FP16 }";

// One colour per id, so the same person keeps the same colour for as long as they are tracked.
static Scalar colorForId(int id)
{
    RNG rng((uint64)id * 0x9E3779B97F4A7C15ULL + 1);
    return Scalar(rng.uniform(64, 256), rng.uniform(64, 256), rng.uniform(64, 256));
}

// yolov8 emits one [1, 84, 8400] tensor: four box numbers then a score per class. The image is
// padded to a square first, so the boxes scale back by the padded length.
static void detect(const Mat& frame, Net& net, int inputSize, double inputScale, bool swapRB,
                   float confThreshold, float nmsThreshold,
                   vector<Rect2d>& boxes, vector<float>& scores, vector<int>& classIds)
{
    boxes.clear();
    scores.clear();
    classIds.clear();

    const int length = max(frame.rows, frame.cols);
    Mat square = Mat::zeros(Size(length, length), frame.type());
    frame.copyTo(square(Rect(0, 0, frame.cols, frame.rows)));
    const double backScale = (double)length / inputSize;

    Mat blob;
    blobFromImage(square, blob, inputScale, Size(inputSize, inputSize), Scalar(), swapRB, false, CV_32F);
    net.setInput(blob);

    vector<Mat> outs;
    net.forward(outs, net.getUnconnectedOutLayersNames());

    Mat out = outs[0].reshape(0, outs[0].size[1]);   // [84, 8400]
    Mat cand;
    transpose(out, cand);                            // [8400, 84]

    vector<Rect2d> raw;
    vector<float> rawScores;
    vector<int> rawClasses;
    for (int i = 0; i < cand.rows; i++)
    {
        double maxScore = 0.0, minScore = 0.0;
        Point maxLoc, minLoc;
        minMaxLoc(cand.row(i).colRange(4, cand.cols), &minScore, &maxScore, &minLoc, &maxLoc);
        if (maxScore < confThreshold || maxLoc.x != 0)
            continue;                                // class 0 is "person" in COCO

        const double cx = cand.at<float>(i, 0);
        const double cy = cand.at<float>(i, 1);
        const double w = cand.at<float>(i, 2);
        const double h = cand.at<float>(i, 3);
        raw.push_back(Rect2d(cx - 0.5 * w, cy - 0.5 * h, w, h));
        rawScores.push_back((float)maxScore);
        rawClasses.push_back(maxLoc.x);
    }

    vector<int> keep;
    NMSBoxes(raw, rawScores, confThreshold, nmsThreshold, keep);
    for (size_t i = 0; i < keep.size(); i++)
    {
        const Rect2d& b = raw[keep[i]];
        boxes.push_back(Rect2d(b.x * backScale, b.y * backScale,
                               b.width * backScale, b.height * backScale));
        scores.push_back(rawScores[keep[i]]);
        classIds.push_back(rawClasses[keep[i]]);
    }
}

// One L2-normalised descriptor per detection. Model is used over blobFromImage because
// setInputScale takes a Scalar, so the per-channel std folds straight in.
static Mat describe(const Mat& frame, const Ptr<Model>& reid, const vector<Rect2d>& boxes)
{
    if (boxes.empty())
        return Mat();

    Mat features;
    for (size_t i = 0; i < boxes.size(); i++)
    {
        Rect roi = Rect((int)boxes[i].x, (int)boxes[i].y,
                        (int)boxes[i].width, (int)boxes[i].height) & Rect(0, 0, frame.cols, frame.rows);
        if (roi.width < 2 || roi.height < 2)
            roi = Rect(0, 0, min(2, frame.cols), min(2, frame.rows));

        vector<Mat> outs;
        reid->predict(frame(roi), outs);
        Mat f = outs[0].reshape(1, 1);
        f.convertTo(f, CV_32F);
        const double n = norm(f);
        if (n > 0)
            f /= n;
        features.push_back(f);
    }
    return features;
}

int main(int argc, char** argv)
{
    CommandLineParser parser(argc, argv, keys);
    parser.about(about);
    if (!parser.has("@alias") || parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    const string modelName = parser.get<String>("@alias");
    const string zooFile = findFile(parser.get<String>("zoo"));
    const string allKeys = keys + genPreprocArguments(modelName, zooFile)
                                + genPreprocArguments(modelName, zooFile, "yolo_");
    parser = CommandLineParser(argc, argv, allKeys);
    parser.about(about);

    const bool useReid = parser.has("reid");
    const int backend = parser.get<int>("backend");
    const int target = parser.get<int>("target");

    // Detector
    const string yoloPath = findModel(parser.get<String>("yolo_model"), parser.get<String>("yolo_sha1"));
    Net yolo = readNetFromONNX(yoloPath);
    yolo.setPreferableBackend((dnn::Backend)backend);
    yolo.setPreferableTarget((dnn::Target)target);
    const int yoloSize = parser.get<int>("yolo_width");
    const double yoloScale = parser.get<double>("yolo_scale");
    const bool yoloSwapRB = parser.get<bool>("yolo_rgb");

    // Appearance, only when asked for
    Ptr<Model> reid;
    if (useReid)
    {
        const string reidPath = findModel(parser.get<String>("model"), parser.get<String>("sha1"));
        reid = makePtr<Model>(readNetFromONNX(reidPath));
        const Scalar mean = parser.get<Scalar>("mean");
        const Scalar std = parser.get<Scalar>("std");
        const double scale = parser.get<double>("scale");
        // (pixel - mean) * scale, per channel, with the std folded into the scale. swapRB swaps
        // channel 0 and 2 of both, so each is given in the model's own RGB order.
        reid->setInputMean(mean);
        reid->setInputScale(Scalar(scale / std[0], scale / std[1], scale / std[2]));
        reid->setInputSize(parser.get<int>("width"), parser.get<int>("height"));
        reid->setInputSwapRB(parser.get<bool>("rgb"));
        reid->setPreferableBackend((dnn::Backend)backend);
        reid->setPreferableTarget((dnn::Target)target);
    }

    MultiTracker::Params params;
    if (useReid)
        params.embeddingWeight = 0.4f;   // appearance helps, but motion still leads
    Ptr<MultiTracker> tracker = MultiTracker::create(params);

    const string inputName = parser.get<String>("input");
    VideoCapture cap;
    if (inputName.empty() || (isdigit(inputName[0]) && inputName.size() == 1))
        cap.open(inputName.empty() ? parser.get<int>("device") : inputName[0] - '0');
    else
        cap.open(findFile(inputName));
    if (!cap.isOpened())
    {
        cerr << "Could not open the input " << inputName << endl;
        return 1;
    }

    const string window = "MULTI OBJECT TRACKING";
    namedWindow(window, WINDOW_NORMAL);
    FontFace fontFace("sans");
    const int stdSize = 15, stdWeight = 400, stdImgSize = 512;

    Mat frame;
    for (;;)
    {
        cap >> frame;
        if (frame.empty())
            break;

        const int fontSize = max(10, stdSize * frame.cols / stdImgSize);
        const int fontWeight = max(200, stdWeight * frame.cols / stdImgSize);

        vector<Rect2d> detBoxes;
        vector<float> detScores;
        vector<int> detClasses;
        detect(frame, yolo, yoloSize, yoloScale, yoloSwapRB,
               parser.get<float>("conf"), parser.get<float>("nms"),
               detBoxes, detScores, detClasses);

        vector<int> ids;
        vector<Rect2d> boxes;
        vector<int> classes;
        if (useReid)
            tracker->update(detBoxes, detScores, detClasses, describe(frame, reid, detBoxes),
                            ids, boxes, classes);
        else
            tracker->update(detBoxes, detScores, detClasses, ids, boxes, classes);

        for (size_t i = 0; i < ids.size(); i++)
        {
            const Scalar color = colorForId(ids[i]);
            const Rect box((int)boxes[i].x, (int)boxes[i].y, (int)boxes[i].width, (int)boxes[i].height);
            rectangle(frame, box, color, 2);

            const string label = format("%d", ids[i]);
            Rect tr = getTextSize(Size(), label, Point(), fontFace, fontSize, fontWeight);
            tr.height += fontSize / 2;
            tr.width += fontSize / 2;
            tr.x = box.x;
            tr.y = max(0, box.y - tr.height);
            rectangle(frame, tr, color, FILLED);
            putText(frame, label, Point(tr.x + fontSize / 4, tr.y + fontSize),
                    Scalar::all(0), fontFace, fontSize, fontWeight);
        }

        const string status = format("tracking %d  (detections %d)%s",
                                     (int)ids.size(), (int)detBoxes.size(),
                                     useReid ? "  [re-id on]" : "");
        putText(frame, status, Point(10, fontSize + 5), Scalar(0, 255, 0),
                fontFace, fontSize, fontWeight);

        imshow(window, frame);
        if (waitKey(30) >= 0)
            break;
    }
    return 0;
}
